Linux X11: dragging out a Firefox tab onto Zed crashes it

### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

- Open Zed and Firefox
- Create two tabs in Firefox with arbitrary contents
- Drag out a tab into a separate window, dropping it over the Zed window
- SIGABRT

Sometimes it takes 3-4 attempts to crash if both windows are on the same monitor. The chance seems to be higher if the applications are on different screens.

### Environment

Zed: v0.155.2 (Zed)
OS: Linux X11 debian unknown
Memory: 62.7 GiB
Architecture: x86_64
GPU: NVIDIA GeForce RTX 3070 Ti || NVIDIA || 535.183.06

Extra info: this is Debian Testing running Gnome.

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

<details><summary>Zed.log</summary><pre>
<!-- Click below this line and paste or drag-and-drop your log-->
2024-10-07T15:31:50.07828683+02:00 [WARN] unhandled capability registration: Registration { id: "d6d6b084-3566-4321-88d5-66514975c1ed", method: "workspace/didChangeWorkspaceFolders", register_options: Some(Object {}) }
2024-10-07T15:31:55.655321643+02:00 [ERROR] {
  "thread": "main",
  "payload": "already borrowed: BorrowMutError",
  "location_data": {
    "file": "crates/gpui/src/platform/linux/x11/client.rs",
    "line": 1365
  },
  "backtrace": [
    "core::panicking::panic_fmt",
    "core::cell::panic_already_borrowed",
    "<gpui::platform::linux::x11::client::X11Client as gpui::platform::linux::platform::LinuxClient>::set_cursor_style",
    "gpui::window::WindowContext::dispatch_event",
    "gpui::window::Window::new::{{closure}}",
    "gpui::platform::linux::x11::window::X11WindowStatePtr::handle_input",
    "gpui::platform::linux::x11::client::X11Client::handle_event",
    "gpui::platform::linux::x11::client::X11Client::process_x11_events",
    "<core::cell::RefCell<calloop::sources::DispatcherInner<S,F>> as calloop::sources::EventDispatcher<Data>>::process_events",
    "<gpui::platform::linux::x11::client::X11Client as gpui::platform::linux::platform::LinuxClient>::run",
    "gpui::platform::linux::platform::<impl gpui::platform::Platform for P>::run",
    "gpui::app::App::run",
    "zed::main",
    "std::sys::backtrace::__rust_begin_short_backtrace",
    "std::rt::lang_start::{{closure}}",
    "std::rt::lang_start_internal",
    "main",
    "__libc_start_main",
    "_start"
  ],
  "app_version": "0.155.2",
  "release_channel": "Zed",
  "os_name": "Linux X11",
  "os_version": "debian unknown",
  "architecture": "x86_64",
  "panicked_on": 1728307915655,
  "system_id": "7db9a251-93ba-454b-84b0-cac63e7be65f",
  "installation_id": "92130564-5cb0-41fb-9b93-b904f9e9295f",
  "session_id": "d8015916-36db-48c9-a0a7-2c6266487971"
}

<!-- Click above this line and paste or drag-and-drop your log--></pre></details>

--- a/crates/gpui/src/platform/linux/x11/client.rs
+++ b/crates/gpui/src/platform/linux/x11/client.rs
@@ -675,11 +675,12 @@ impl X11Client {
                         }
                     }
                 } else if event.type_ == state.atoms.XdndLeave {
-                    window.handle_input(PlatformInput::FileDrop(FileDropEvent::Pending {
-                        position: state.xdnd_state.position,
-                    }));
+                    let position = state.xdnd_state.position;
+                    drop(state);
+                    window
+                        .handle_input(PlatformInput::FileDrop(FileDropEvent::Pending { position }));
                     window.handle_input(PlatformInput::FileDrop(FileDropEvent::Exited {}));
-                    state.xdnd_state = Xdnd::default();
+                    self.0.borrow_mut().xdnd_state = Xdnd::default();
                 } else if event.type_ == state.atoms.XdndPosition {
                     if let Ok(pos) = state
                         .xcb_connection
@@ -709,9 +710,10 @@ impl X11Client {
                         state.xdnd_state.other_window,
                         arg4,
                     );
-                    window.handle_input(PlatformInput::FileDrop(FileDropEvent::Pending {
-                        position: state.xdnd_state.position,
-                    }));
+                    let position = state.xdnd_state.position;
+                    drop(state);
+                    window
+                        .handle_input(PlatformInput::FileDrop(FileDropEvent::Pending { position }));
                 } else if event.type_ == state.atoms.XdndDrop {
                     xdnd_send_finished(
                         &state.xcb_connection,
@@ -719,10 +721,11 @@ impl X11Client {
                         event.window,
                         state.xdnd_state.other_window,
                     );
-                    window.handle_input(PlatformInput::FileDrop(FileDropEvent::Submit {
-                        position: state.xdnd_state.position,
-                    }));
-                    state.xdnd_state = Xdnd::default();
+                    let position = state.xdnd_state.position;
+                    drop(state);
+                    window
+                        .handle_input(PlatformInput::FileDrop(FileDropEvent::Submit { position }));
+                    self.0.borrow_mut().xdnd_state = Xdnd::default();
                 }
             }
             Event::SelectionNotify(event) => {
@@ -751,8 +754,9 @@ impl X11Client {
                                 position: state.xdnd_state.position,
                                 paths: crate::ExternalPaths(paths),
                             });
+                            drop(state);
                             window.handle_input(input);
-                            state.xdnd_state.retrieved = true;
+                            self.0.borrow_mut().xdnd_state.retrieved = true;
                         }
                         Err(_) => {}
                     }

